core/search/service.py

```diff
@@ -0,0 +1,330 @@
+import logging
+import re
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Iterable, List, Optional
+
+import opensearchpy.helpers
+from opensearch_dsl import MultiSearch, Search
+from opensearchpy import NotFoundError, OpenSearch, RequestError
+
+from core.search.revision import SearchSchemaRevision
+
+
+@dataclass
+class SearchWritePointer:
+    """The 'write' pointer; the pointer that will be used to populate an index with search documents."""
+
+    base_name: str
+    version: int
+
+    @property
+    def name(self) -> str:
+        return f"{self.base_name}-search-write"
+
+    @property
+    def target_name(self) -> str:
+        return f"{self.base_name}-v{self.version}"
+
+
+class SearchServiceException(Exception):
+    """The type of exceptions raised by the search service."""
+
+    def __init__(self, message: str):
+        super().__init__(message)
+
+
+@dataclass
+class SearchServiceFailedDocument:
+    """An error indicating that a document failed to index."""
+
+    id: int
+    error_message: str
+    error_status: int
+    error_exception: str
+
+    @classmethod
+    def from_bulk_error(cls, error: dict):
+        """Transform an error dictionary returned from opensearchpy's bulk API to a typed error"""
+        if error.get("index"):
+            error_indexed = error["index"]
+            error_id = int(error_indexed["_id"])
+            error_status = error_indexed["status"]
+            error_reason = error_indexed["error"]["reason"]
+            return SearchServiceFailedDocument(
+                id=error_id,
+                error_message=error_reason,
+                error_status=error_status,
+                error_exception="<unavailable>",
+            )
+        else:
+            raise NotImplementedError
```

Comment:
Minor: raise NotImplementedError`()`

Reply:
:+1: 

Oversight. Didn't mean to leave this one in.